# Easy Visa

### Importing Libraries

In [2]:
import warnings

warnings.filterwarnings('ignore')

# data manipulation libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# modeling libraries
from sklearn.model_selection import train_test_split, GridSearchCV

from xgboost import XGBClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier

from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, f1_score, roc_auc_score, roc_curve, \
    auc, precision_recall_curve, recall_score, precision_score

### Importing Data

In [8]:
# original data
visa = pd.read_csv('EasyVisa.csv')

# copying the data to avoid any changes in the original data
data = visa.copy()

### Data Information

In [23]:
data

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25480 rows x 12 columns]

In [24]:
data.info()
# data is full and there is no missing value

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25480 entries, 0 to 25479
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   case_id                25480 non-null  object 
 1   continent              25480 non-null  object 
 2   education_of_employee  25480 non-null  object 
 3   has_job_experience     25480 non-null  object 
 4   requires_job_training  25480 non-null  object 
 5   no_of_employees        25480 non-null  int64  
 6   yr_of_estab            25480 non-null  int64  
 7   region_of_employment   25480 non-null  object 
 8   prevailing_wage        25480 non-null  float64
 9   unit_of_wage           25480 non-null  object 
 10  full_time_position     25480 non-null  object 
 11  case_status            25480 non-null  object 
dtypes: float64(1), int64(2), object(9)
memory usage: 2.3+ MB


In [25]:
print('There is {} rows, and {} columns in the data'.format(data.shape[0], data.shape[1]))

There is 25480 rows, and 12 columns in the data


In [26]:
print('The columns are:')
data.columns

The columns are:


Index(['case_id', 'continent', 'education_of_employee', 'has_job_experience',
       'requires_job_training', 'no_of_employees', 'yr_of_estab',
       'region_of_employment', 'prevailing_wage', 'unit_of_wage',
       'full_time_position', 'case_status'],
      dtype='object')

In [27]:
print('There are {} duplicated rows in the data.'.format(data.duplicated().sum()))

There are 0 duplicated rows in the data.


### Data Exploration

In [29]:
data.describe().T

count          mean           std        min       25%  \
no_of_employees  25480.0   5667.043210  22877.928848   -26.0000   1022.00   
yr_of_estab      25480.0   1979.409929     42.366929  1800.0000   1976.00   
prevailing_wage  25480.0  74455.814592  52815.942327     2.1367  34015.48   

                      50%          75%        max  
no_of_employees   2109.00    3504.0000  602069.00  
yr_of_estab       1997.00    2005.0000    2016.00  
prevailing_wage  70308.21  107735.5125  319210.27

In [28]:
data.describe(include='all').T

count unique         top   freq          mean  \
case_id                  25480  25480      EZYV01      1           NaN   
continent                25480      6        Asia  16861           NaN   
education_of_employee    25480      4  Bachelor's  10234           NaN   
has_job_experience       25480      2           Y  14802           NaN   
requires_job_training    25480      2           N  22525           NaN   
no_of_employees        25480.0    NaN         NaN    NaN    5667.04321   
yr_of_estab            25480.0    NaN         NaN    NaN   1979.409929   
region_of_employment     25480      5   Northeast   7195           NaN   
prevailing_wage        25480.0    NaN         NaN    NaN  74455.814592   
unit_of_wage             25480      4        Year  22962           NaN   
full_time_position       25480      2           Y  22773           NaN   
case_status              25480      2   Certified  17018           NaN   

                                std     min       25%       50%          75%  \
case_id                         NaN     NaN       NaN       NaN          NaN   
continent                       NaN     NaN       NaN       NaN          NaN   
education_of_employee           NaN     NaN       NaN       NaN          NaN   
has_job_experience              NaN     NaN       NaN       NaN          NaN   
requires_job_training           NaN     NaN       NaN       NaN          NaN   
no_of_employees        22877.928848   -26.0    1022.0    2109.0       3504.0   
yr_of_estab               42.366929  1800.0    1976.0    1997.0       2005.0   
region_of_employment            NaN     NaN       NaN       NaN          NaN   
prevailing_wage        52815.942327  2.1367  34015.48  70308.21  107735.5125   
unit_of_wage                    NaN     NaN       NaN       NaN          NaN   
full_time_position              NaN     NaN       NaN       NaN          NaN   
case_status                     NaN     NaN       NaN       NaN          NaN   

                             max  
case_id                      NaN  
continent                    NaN  
education_of_employee        NaN  
has_job_experience           NaN  
requires_job_training        NaN  
no_of_employees         602069.0  
yr_of_estab               2016.0  
region_of_employment         NaN  
prevailing_wage        319210.27  
unit_of_wage                 NaN  
full_time_position           NaN  
case_status                  NaN

**Observations:**
* The range of the number of employees in a company is huge. There are some anomalies in the data as we can see that the minimum number of employees is equal to -26, which is not possible. We will have to fix this.
* The year of establishment of companies ranges from 1800 to 2016, which seems fine.
* The average prevailing wage is 74455.81. There's also a very huge difference in 75th percentile and maximum value which indicates there might be outliers present in this column.